### PR TITLE
Added MapToCurve interface

### PR DESCRIPTION
**Pull request by nikhilvanjani**

### CODE REVIEW

File: src/map_to_curve.rs [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+            let mut tmp = PtT::osswu_map(p1);
+            tmp.add_assign(&PtT::osswu_map(p2));
+            tmp
+        };
+        p.isogeny_map();
+        p.clear_h();
+        p
+	}
+
+}
```
Comment (kwantam) [inline src/map_to_curve.rs:36]:
```suggestion
}
```

(nice to have a trailing newline imo)

**Review by kwantam — CHANGES_REQUESTED** [review]
The most significant bit is that this looks good! :+1:

But I think maybe Clippy suggested some changes that should be rolled back, and there's a small bit of cleanup to do here and there. See inline comments.

File: src/lib.rs [hunk 2]
```diff
@@ -362,7 +363,7 @@ pub trait EncodedPoint:
 
     /// Converts an `EncodedPoint` into a `CurveAffine` element,
     /// if the encoding represents a valid element.
-    fn into_affine(&self) -> Result<Self::Affine, GroupDecodingError>;
+    fn into_affine(self) -> Result<Self::Affine, GroupDecodingError>;
```
[the remark this answers is not in the document]
Comment (nikhilvanjani) [reply src/lib.rs:366]: Yes, this was in response to clippy warnings. Sure, I will make the breaking change.

[the remark this answers is not in the document]
Comment (nikhilvanjani) [reply src/lib.rs:366]: the following was the error. Not sure, what must I allow in this case.

error: methods called `into_*` usually take `self` by value
   --> src/lib.rs:211:20
    |
211 |     fn into_affine(&self) -> Self::Affine;
    |                    ^^^^^
    |
note: the lint level is defined here
   --> src/lib.rs:4:44
    |
4   | #![cfg_attr(feature = "cargo-clippy", deny(warnings))]
    |                                            ^^^^^^^^
    = note: `#[deny(clippy::wrong_self_convention)]` implied by `#[deny(warnings)]`
    = help: consider choosing a less ambiguous name
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#wrong_self_convention

File: src/lib.rs [hunk 3]
```diff
@@ -419,7 +420,7 @@ impl fmt::Display for GroupDecodingError {
             GroupDecodingError::CoordinateDecodingError(description, ref err) => {
                 write!(f, "{} decoding error: {}", description, err)
             }
-            _ => write!(f, "{}", self.to_string()),
+            _ => write!(f, "{:?}", self),
```
[the remark this answers is not in the document]
Comment (nikhilvanjani) [reply src/lib.rs:423]: Yes the following clippy warning -
====
error: using `to_string` in `fmt::Display` implementation might lead to infinite recursion
   --> src/lib.rs:423:34
    |
423 |             _ => write!(f, "{}", self.to_string()),
    |                                  ^^^^^^^^^^^^^^^^
    |
    = note: `#[deny(clippy::to_string_in_display)]` on by default
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#to_string_in_display
====
I tried to follow suggestions from the last help link above and it suggested to do `write!(f, "{}", self.0)`. But this resulted in the error `error[E0609]: no field `0` on type `&GroupDecodingError``. So, I moved to `write!(f, "{:?}", self)`. I can do something else if there's any suggestion.

File: src/map_to_curve.rs [hunk 4]
```diff
@@ -0,0 +1,36 @@
+use bls12_381::{ClearH, IsogenyMap, OSSWUMap};
+use CurveProjective;
+
+pub trait MapToCurve<PtT>
+where
+	PtT: CurveProjective,
+{
+	fn map_one_field_element_to_curve(p1: &PtT::Base) -> PtT;
+	fn map_two_field_elements_to_curve(p1: &PtT::Base, p2: &PtT::Base) -> PtT;
```
[the remark this answers is not in the document]
Comment (nikhilvanjani) [reply src/map_to_curve.rs:9]: Sure, I can change this.

File: src/map_to_curve.rs [hunk 5]
```diff
@@ -0,0 +1,36 @@
+use bls12_381::{ClearH, IsogenyMap, OSSWUMap};
```
[the remark this answers is not in the document]
Comment (nikhilvanjani) [reply src/map_to_curve.rs:1]: done.

File: src/map_to_curve.rs [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+	fn map_one_field_element_to_curve(p1: &PtT::Base) -> PtT {
+        println!("map_one_field_element_to_curve called");
+        let mut p = PtT::osswu_map(p1);
+        p.isogeny_map();
+        p.clear_h();
+        p
+	}
+
+	fn map_two_field_elements_to_curve(p1: &PtT::Base, p2: &PtT::Base) -> PtT {
+        println!("map_two_field_elements_to_curve called");
```
[the remark this answers is not in the document]
Comment (nikhilvanjani) [reply src/map_to_curve.rs:25]: done.

File: src/map_to_curve.rs [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+	fn map_one_field_element_to_curve(p1: &PtT::Base) -> PtT;
+	fn map_two_field_elements_to_curve(p1: &PtT::Base, p2: &PtT::Base) -> PtT;
+}
+
+impl<PtT> MapToCurve<PtT> for PtT
+where
+    PtT: ClearH + IsogenyMap + OSSWUMap,
+{
+	fn map_one_field_element_to_curve(p1: &PtT::Base) -> PtT {
+        println!("map_one_field_element_to_curve called");
```
[the remark this answers is not in the document]
Comment (nikhilvanjani) [reply src/map_to_curve.rs:17]: done.

File: src/hash_to_curve.rs [hunk 8]
```diff
@@ -27,24 +28,28 @@ where
...
-        p
+        // let mut p = {
+        //     let u = hash_to_field::<CoordT<PtT>, X>(msg.as_ref(), dst.as_ref(), 2);
+        //     let mut tmp = PtT::osswu_map(&u[0]);
+        //     tmp.add_assign(&PtT::osswu_map(&u[1]));
+        //     tmp
+        // };
+        // p.isogeny_map();
+        // p.clear_h();
+        // p
```
[the remark this answers is not in the document]
Comment (nikhilvanjani) [reply src/hash_to_curve.rs:39]: done.

File: src/hash_to_curve.rs [hunk 9]
```diff
@@ -27,24 +28,28 @@ where
...
-        p.isogeny_map();
-        p.clear_h();
-        p
+        // let mut p = {
+        //     let u = hash_to_field::<CoordT<PtT>, X>(msg.as_ref(), dst.as_ref(), 1);
+        //     PtT::osswu_map(&u[0])
+        // };
+        // p.isogeny_map();
+        // p.clear_h();
+        // p
```
[the remark this answers is not in the document]
Comment (nikhilvanjani) [reply src/hash_to_curve.rs:51]: done.

File: src/map_to_curve.rs (hunk 1 above)
Comment (nikhilvanjani) [reply src/map_to_curve.rs:36]: done.

File: src/map_to_curve.rs [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+        let mut p = PtT::osswu_map(p1);
+        p.isogeny_map();
+        p.clear_h();
+        p
```
[the remark this answers is not in the document]
Comment (nikhilvanjani) [reply src/map_to_curve.rs:21]: Sounds reasonable to do. An optimization could be-
`debug_assert!(p.into_affine().in_subgroup());`
based on the following code in g1.rs:319:
`
    impl SubgroupCheck for G1Affine {
        fn in_subgroup(&self) -> bool {
            self.is_on_curve() && self.is_in_correct_subgroup_assuming_on_curve()
        }
    }
`

File: src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (kwantam) [reply src/lib.rs]: :+1:

File: src/lib.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (kwantam) [reply src/lib.rs:425]: Got it. Seems OK then.

File: src/map_to_curve.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (kwantam) [reply src/map_to_curve.rs:25]: Excellent, that sounds exactly right.

[the remark this answers is not in the document]
Comment (kwantam) [reply src/map_to_curve.rs:25]: So I guess the new version would be:

```suggestion
        debug_assert!(p.into_affine().in_subgroup());
        p
```

right?

File: src/map_to_curve.rs [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+	fn map_two_field_elements_to_curve(p1: &PtT::Base, p2: &PtT::Base) -> PtT {
+        println!("map_two_field_elements_to_curve called");
+        let mut p = {
+            let mut tmp = PtT::osswu_map(p1);
+            tmp.add_assign(&PtT::osswu_map(p2));
+            tmp
+        };
+        p.isogeny_map();
+        p.clear_h();
+        p
```
[the remark this answers is not in the document]
Comment (kwantam) [reply src/map_to_curve.rs:36]: As above,

```suggestion
        debug_assert!(p.into_affine().in_subgroup());
        p
```

File: src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (kwantam) [reply src/lib.rs]: The fix is to make the declaration the following:

```
        #[allow(clippy::wrong_self_convention)]
        fn into_affine(&self) -> Result<Self::Affine, GroupDecodingError>;
```

File: src/lib.rs [hunk 12]
```diff
@@ -371,7 +372,7 @@ pub trait EncodedPoint:
     ///
     /// If the encoding is invalid, this can break API invariants,
     /// so caution is strongly encouraged.
-    fn into_affine_unchecked(&self) -> Result<Self::Affine, GroupDecodingError>;
+    fn into_affine_unchecked(self) -> Result<Self::Affine, GroupDecodingError>;
```
[the remark this answers is not in the document]
Comment (kwantam) [reply src/lib.rs]: Also here:

```rust
        #[allow(clippy::wrong_self_convention)]
        fn into_affine_unchecked(&self) -> Result<Self::Affine, GroupDecodingError>;
```

**Review by kwantam — CHANGES_REQUESTED** [review]
Cool! A couple further comments. Adding the `#[allow(...)]`s will make the tests pass.

File: src/lib.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (nikhilvanjani) [reply src/lib.rs]: done.

File: src/lib.rs [hunk 13]
```diff
@@ -13,6 +13,8 @@
 // Force public structures to implement Debug
 #![deny(missing_debug_implementations)]
 
+#![allow(clippy::wrong_self_convention)]
+
```
[the remark this answers is not in the document]
Comment (kwantam) [reply src/lib.rs]: :+1:

File: src/lib.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (kwantam) [reply src/lib.rs]: :+1:

**Review by kwantam — APPROVED** [review]
Looks good to me! One last `cargo fmt` and `cargo clippy`, maybe?

**Pull request merged**